v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts

```diff
@@ -0,0 +1,70 @@
+import type { NewTaskActionFunction } from "@ignored/hardhat-vnext-core/types/tasks";
+import type { REPLServer } from "node:repl";
+
+import path from "node:path";
+import repl from "node:repl";
+
+import { getCacheDir } from "@ignored/hardhat-vnext-core/global-dir";
+import debug from "debug";
+
+const log = debug("hardhat:core:tasks:console");
+
+interface ConsoleActionArguments {
+  commands: string[];
+  history: string;
+  noCompile: boolean;
+  // We accept ReplOptions as an argument to allow tests overriding the IO streams
+  options?: repl.ReplOptions;
+}
+
+const consoleAction: NewTaskActionFunction<ConsoleActionArguments> = async (
+  { commands, history, noCompile, options },
+  hre,
+) => {
+  // Resolve the history path if it is not empty
+  let historyPath: string | undefined;
+  if (history !== "") {
+    const globalCacheDir = await getCacheDir();
+    historyPath = path.isAbsolute(history)
+      ? history
+      : path.resolve(globalCacheDir, history);
+  }
+
+  // If noCompile is false, run the compile task first
+  if (!noCompile) {
+    // todo: run compile task
+  }
+
+  return new Promise<REPLServer>(async (resolve) => {
+    // Start a new REPL server with the default options
+    const replServer = repl.start(options);
```

Comment:
Unfortunately, the only option to create an instance of REPL is to start it immediately. This is because the constructor of the REPL class is private. Ideally, we would only start it after the rest of the setup is complete.

Reply:
It's crazy, but unfortunately, not surprising.